go/merkletree2/position.go

```diff
@@ -96,18 +96,72 @@ func (t *Tree) getDeepestPositionForKey(k Key) (*Position, error) {
 	return &p, nil
 }
 
-func (t *Tree) getSiblingPositionsOnPathToKey(k Key) ([][]Position, error) {
+// getSiblingPositionsOnPathToKey returns a slice of positions, in descending
+// order by level (siblings farther from the root come first) and in
+// lexicographic order within each level.
+func (t *TreeConfig) getSiblingPositionsOnPathToKey(k Key) ([]Position, error) {
 	p, err := t.getDeepestPositionForKey(k)
 	if err != nil {
 		return nil, err
 	}
-	maxPathLength := ((*big.Int)(p).BitLen() - 1) / int(t.cfg.bitsPerIndex)
-	positions := make([][]Position, maxPathLength)
+	maxPathLength := t.keysByteLength * 8 / int(t.bitsPerIndex)
+	positions := make([]Position, 0, maxPathLength*(t.childrenPerNode-1))
 	root := t.getRootPosition()
+	var sibs []Position
 	for i := 0; !p.equals(root); {
-		positions[i], p = t.getAllSiblings(p)
+		sibs, p = t.getAllSiblings(p)
+		positions = append(positions, sibs...)
 		i++
 	}
 
 	return positions, nil
 }
+
+// getLevel returns the level of p. The root is at level 0, and each node has
+// level 1 higher than its parent.
+func (t *TreeConfig) getLevel(p *Position) int {
+	return ((*big.Int)(p).BitLen() - 1) / int(t.bitsPerIndex)
+}
+
+// getParentAtLevel returns nil if p is at a level lower than `level`. The root
+// is at level 0, and each node has level 1 higher than its parent.
+func (t *TreeConfig) getParentAtLevel(p *Position, level uint) *Position {
+	shift := (*big.Int)(p).BitLen() - 1 - int(t.bitsPerIndex)*int(level)
+	if (*big.Int)(p).BitLen() < 2 || shift < 0 {
+		return nil
+	}
+
+	var f big.Int
+	f.Rsh((*big.Int)(p), uint(shift))
+
+	return (*Position)(&f)
+}
+
+// positionToChildIndexes returns the list of childIndexes to navigate from the
+// root to p (in reverse order).
+func (t *TreeConfig) positionToChildIndexes(p *Position) (path []ChildIndex) {
+	path = make([]ChildIndex, ((*big.Int)(p).BitLen()-1)/int(t.bitsPerIndex))
```

Comment:
`((*big.Int)(p).BitLen()-1)/int(t.bitsPerIndex)` this seems like a common expr, maybe it could be factored out